### DISCUSSION TITLE
Access Collapsible context from Collapsible.Content wrapper

### DISCUSSION THREAD

**Discussion by kuus** [radix-ui/primitives#1897] [Help]
Hi guys, I am trying to understand how to access the context of the Collapsible, I just need the `open` state when wrapping a `Collapsible.Content` component into a custom component, is it doable? it seems there is no hook exposed for that..
thanks in advance!

**Comment by joaom00** [comment]
Hey @kuus, you can control the `Collapsible` passing `open` and `onOpenChange`

**Reply by kuus** [reply]
Hi @joaom00 thanks for your reply. Yes sure I can make it controlled but what I wanted was to keep the uncontrolled version, basically I wanted to change some fundamental behaviours of the primitive  component by component, so it made more sense to just rewrite it. What could have helped mel would have been to have `useCollapsibleContext` exposed from the radix primtive, but in the end is so little code that it worth just rewriting it copy pasting the [Collapsible source](https://github.com/radix-ui/primitives/blob/main/packages/react/collapsible/src/Collapsible.tsx) as a starting point. Bonus point of radix to me is in fact the very scoped packaging organization, which helps in extending their primitives without exploding the bundle size and without re-writing a lot of helpers

**Reply by joaom00** [reply]
What are you trying to achieve with access to `open` state?

**Reply by kuus** [reply]
I am building/using a disclosure component inspired by this https://google.github.io/closure-library/source/closure/goog/demos/ (`zippy` demo),
plus an optionable collapse/expand behaviour from bottom
It does not animate height but it rather works on marginTop/marginBottom, and it needs some special stuff on the various basic UI elements to work correctly

**Reply by benoitgrelard** [reply]
This seems like it would be achievable with our primitive no? What specifically doesn't work here?

**Reply by kuus** [reply]
Hi @benoitgrelard how can I extend individually `Collapsible.Root` and `Collapsible.Content` to let's say my implementation `MyCollapsible.Root` and `MyCollapsible.Content` when inside the latter I'd need to access the `open` state from context? It doesn't seem possible right?

**Reply by benoitgrelard** [reply]
I'm not following what you mean by:

> how can I extend individually `Collapsible.Root` and `Collapsible.Content`

Wanna create a sandbox with an example of what you're trying to achieve?

**Reply by kuus** [reply]
When I'll have a bit of time I ll do it!
What I mean is simply something like this:


```tsx
import * as Collapsible from "@radix-ui/react-collapsible";

type DisclosureRootProps = Collapsible.CollapsibleProps & {
  myProp?: boolean;
};

const DisclosureRoot = (props: DisclosureRootProps) => {
  const { myProp, ...rest } = props;
  return (
    <Collapsible.Root
      className={myProp ? "a" : "b"}
      {...rest}
    />
  );
};

type DisclosureContentProps = Collapsible.CollapsibleContentProps & {
  reverse?: boolean;
};

const DisclosureContent = (props: DisclosureContentProps) => {
  const { reverse, ...rest } = props;
  const open = true; // I have no context access here

  return (
    <Collapsible.Content
      className={reverse ? "a" : "b"}
      style={{
        marginTop: reverse && open ? "-1em" : "1em",
      }}
      {...rest}
    />
  );
};

export const Disclosure = Object.assign(
  {},
  {
    Root: DisclosureRoot,
    Content: DisclosureContent,
    Trigger: Collapsible.Trigger,
  }
);

export default Disclosure;



// Usage

const MyDisclosure = () => {
  return (
    <Disclosure.Root>
      <div>
        <p>Something and then</p>
        <Disclosure.Trigger asChild>
          <button>My trigger</button>
        </Disclosure.Trigger>
      </div>
      <Disclosure.Content>
        here something collapsible
      </Disclosure.Content>
    </Disclosure.Root>
  )
}
```

**Reply by benoitgrelard** [reply]
The technicality of it, yes you would need to control the prop if you specifically needed it here.
Still not really a good reason to have to re-write the component from scratch though really.

But my question was rather, what effect specifically are you trying to achieve here?
I am trying to see if there would be a way to fulfil what you are after, without even needing to control alltogether.

**Reply by joaom00** [reply]
The only way I know of you can access the `open` state is to have the component be controlled and pass as props, but since you are extending them individually you need to create a react context and pass through it. Maybe @benoitgrelard knows how to do this better

**Reply by kuus** [reply]
@joaom00  yes creating another context would work but it feels like a duplication. The effect I am after is the same as zippy from closure library I linked before, which I kinda replicated in vanillajs yeaaars ago here https://kuus.github.io/zacc/
for this effect to work I need to control some styling like `overflow` and margins at the component level (Root and Content), plus I have some other options to add to better support SSR in a flexible way (mounting / unmounting / initial state etc.)

**Reply by benoitgrelard** [reply]
Here's a demo of the effect using CSS animations: https://codesandbox.io/s/accordion-sliding-under-animation-74uyiz?file=/styles.css:1693-1959

**Reply by kuus** [reply]
hi @benoitgrelard that is nice thanks, what is not ideal is to have `overflow: hidden` on the `.AccordionContent` as that would cut off stuff within the content. That CSS rule should be applied only just before the closing animation starts and removed when the opening animation ends.

Honestly I don't think I can achieve everything I would like without extending the primitives as in my previous comment's sample code.. Other things are in fact that the transition time should be controlled by a prop and that the animation should work also in reverse (using marginBottom) when a `reverse` prop is passed

**Reply by benoitgrelard** [reply]
> what is not ideal is to have overflow: hidden on the .AccordionContent as that would cut off stuff within the content.

Not sure what content would be cut-off?

> Honestly I don't think I can achieve everything I would like without extending the primitives as in my previous comment's sample code.. Other things are in fact that the transition time should be controlled by a prop and that the animation should work also in reverse (using marginBottom) when a reverse prop is passed

No worries, I'm just giving you solutions. At the end of the day if you don't feel the primitive is useful to you, you don't have to use it!

**Reply by kuus** [reply]
> Not sure what content would be cut-off?

anyhting absolutely positioned within the AccordionContent that exceeds its boundary for instance some buttons or UI elements overlapping the edges, or even some wide shadows within the AccordionContent

>No worries, I'm just giving you solutions. At the end of the day if you don't feel the primitive is useful to you, you don't have to 
use it!

of course :) and thanks for that! I love using radix ui so far and I appreciate its design and extensibility

**Reply by kuus** [reply]
sorry @benoitgrelard just to be sure, so you do not intend to expose the use context hooks like `useCollapsibleContext`, correct?

**Reply by benoitgrelard** [reply]
We currently have no plans of doing that.
